File: python/flintdb/schema.py
class Schema:
    """Manages Schema definition and enforcing.

    Attributes:
        _schema (dict): The defined schema.
        _has_encrypted_columns (bool): To check for encrypted columns presence.
    """
    def __init__(self) -> None:
        """Initializes Schema object."""
        self._schema = {}
        self._has_encrypted_columns = False

    def __len__(self) -> int:
        """Returns the count of defined schema."""
        return len(self._schema)

    def add(self, column_name: str, column_type: str, **kwargs) -> "Self":
        """Defines new column schema.

        Args:
            column_name: The column name.
            column_type: The column data type.
            **kwargs: Arbitrary keyword argument.
                The arguments can include:
                    encrypted (bool): To encrypt column value.
                    required (bool): To make column not nullable.
                    enum_values (list): The enum type values.

        Raises:
            ValueError: If column type in not supported. Or if invalid enum values given.
            TypeError: If enum values are not list

        Returns:
            Schema instance.
        """
        if column_type not in Schema.valid_types().keys():
            raise ValueError(f"Unsupported type for column \"{column_name}\"")

        if encrypted := bool(kwargs.get("encrypted")):
            self._has_encrypted_columns = True

        self._schema[column_name] = {
            "type": column_type,
            "encrypted": encrypted,
            "required": bool(kwargs.get("required"))
        }

        if "enum" == column_type:
            enum_values = kwargs.get("enum_values")
            if not isinstance(enum_values, list):
                raise TypeError("ENUM values must be of type list")
            elif not enum_values:
                raise ValueError("ENUM values cannot be empty")

            enum_type = {type(value) for value in enum_values}
            if len(enum_type) > 1:
                raise ValueError("ENUM values must be of the same type")

            self._schema[column_name]["enum_values"] = list(set(enum_values))

        return self

    def get(self, column_name: str) -> dict | None:
        """Retrieves column schema.

        Args:
            column_name: The column name.

        Returns:
            Schema definition for column. Or None if column is not defined.
        """
        if column_name not in self._schema:
            return None

        return self._schema[column_name]

    def valid(self, column_name: str, value: "Any") -> bool:
        """Returns whether value is valid data for column.

        Args:
            column_name: The column name.
            value: The value.
        """
        data = self.get(column_name)
        if not data:
            return True
        elif not data["required"] and value is None:
            return True
        elif "enum" == data["type"] and value not in data["enum_values"]:
            return False

        valid = False
        valid_types = Schema.valid_types().get(data["type"])
        for types in valid_types:
            if isinstance(value, types):
                valid = True

        return valid

    def remove(self, column_name: str) -> None:
        """Removes column from schema definition.

        Args:
            column_name: The column name.
        """
        if column_name in self._schema:
            del self._schema[column_name]

    def sorted_schema(self) -> dict:
        """Returns sorted schema definition."""
        return dict(sorted(self._schema.items()))

    def has_encrypted_columns(self) -> bool:
        """Returns whether schema has encrypted column(s)"""
        return self._has_encrypted_columns

    @staticmethod
    def valid_types() -> dict:
        """Returns supported columns data types."""
        return {
            "@list": (list,),
            "@bool": (bool,),
            "@enum": (list, bool, float, int, dict, str),
            "@float": (float,),
            "@integer": (int,),
            "@numeric": (str, int, float),
            "@object": (dict,),
            "@text": (str,)
        }

File: php/flintdb/Schema.php
<?php

namespace FlintDB;

use Countable;

class Schema implements Countable {
  /**
   * Valid data types
   *
   * @var array
   */
  private static array $types = array(
    '@bool'   => 'is_bool',
    '@enum'   => 'in_array',
    '@float'  => 'is_float',
    '@int'    => 'is_integer',
    '@list'   => 'array_is_list',
    '@number' => 'is_numeric',
    '@object' => 'is_array',
    '@text'   => 'is_string'
  );
  
  /**
   * Schema container
   *
   * @var array
   */
  private array $schema;
  
  /**
   * Constructor
   *
   * @param array $schema
   */
  public function __construct( array $schema = [] ) {
    $this->schema = $schema;
  }
  
  /**
   * Adds column
   *
   * @param string $column
   * @param string $type
   * @param bool $required
   * @param bool $encrypted
   * @param mixed[] $options
   * @return self
   * @throws DomainException
   */
  public function add(
                       string $column,
                       string $type,
                       bool $required = false,
                       bool $encrypted = false,
                       mixed ...$options
                     ): self {
    if ( ! isset( self::$types[ $type ] ) ) {
      throw new DomainException( 'Unsupported type for column: ' . $column );
    }
    
    $data = array();
    $data[ 'callback' ] = self::$types[ $type ];
    switch ( $type ) {
      case '@enum':
        if (
             ! isset( $options[ 'enum_values' ] ) ||
             ! is_array( $options[ 'enum_values' ] ) ||
               count( $options[ 'enum_values' ] ) < 1
           ) {
          throw new DomainException( 'Enum values must be an array with one or more elements' );
        }
        
        $data[ 'args' ] = array();
        $data[ 'args' ][] = array_values( $data[ 'enum_values' ] );
        $data[ 'args' ][] = true;
        break;
      case '@bool':
      case '@float':
      case '@int':
      case '@list':
      case '@number':
      case '@object':
      case '@text':
        $data[ 'args' ] = array();
        break;
      default:
        $data[ 'args' ] = (array) ( $options[ 'args' ] ?? [] );
    }
    
    $data[ 'required' ] = $required;
    $data[ 'encrypted' ] = $encrypted;
    $this->schema[ $column ] = $data;
    return $this;
  }
  
  /**
   * Returns column data
   *
   * @param string $column
   * @return array|null
   */
  public function get( string $column ): ?array {
    return $this->schema[ $column ] ?? null;
  }
  
  /**
   * Returns whether data is valid for column
   *
   * @param string $column
   * @param mixed $value
   * @return bool
   */
  public function valid( string $column, mixed $value ): bool {
    try {
      $data = $this->get( $column );
      if ( ! $data ) {
        return true;
      }
      
      $callback = $data[ 'callback' ];
      if ( ! is_callable( $callback ) ) {
        return false;
      }
      
      elseif ( ! $data[ 'required' ] && is_null( $value ) ) {
        return true;
      }
      
      return ( true === $callback( $value, ...$data[ 'args' ] ) );
    } catch ( Throwable $err ) {
      return false;
    }
  }
  
  /**
   * Removes column
   *
   * @param string $column
   * @return self
   */
  public function remove( string $column ): self {
    unset( $this->schema[ $column ] );
    return $this;
  }
  
  /**
   * Checks for encrypted columns
   *
   * @return bool
   */
  public function has_encrypted_columns(): bool {
    foreach ( $this->schema as $column ) {
      if ( $column[ 'encrypted' ] ) {
        return true;
      }
    }
    
    return false;
  }
  
  /**
   * Returns sorted schema
   *
   * @return array
   */
  public function sorted_schema(): array {
    ksort( $this->schema );
    return $this->schema;
  }
  
  /**
   * Returns number of columns
   *
   * @return int
   */
  public function count(): int {
    return count( $this->schema );
  }
}
